nevergrad/optimization/base.py

```diff
@@ -433,19 +433,29 @@ def ask(self) -> p.Parameter:
         current_num_ask = self.num_ask
         # tentatives if a cheap constraint is available
         # TODO: this should be replaced by an optimization algorithm.
-        max_trials = self._constraints_manager.max_trials
+        max_trials = max(
+            1,
+            self._constraints_manager.max_trials // 2
+            if self.budget is None
+            else (self.budget - self.num_ask),
```

Comment:
why `self.budget -  self.num_ask`? right now this does not make sense to me, and this seems unexpected

Reply:
If the optimizer has a maximum budget, we want the budget we use to be smaller than the remaining budget before the optimizer crashes, hence self.budget - self.num_ask.

Reply:
so you probably want to do some kind of mean between self._constraints_manager.max_trials and `self.budget - self.num_ask` so that:
you don't go beyond max_trials, and you dont exhaust more than half the remaining budget of the optimizer

Reply:
min(self._constraints_manager.max_trials, self.budget - self.num_ask)   could make sense.
Eventually all that stuff is going to be equipped with parameters for combining the auxiliary solver and the simple rejection of mutations.
Besides adding the famous constraint learning tool for expensive constraints that would be useful for some of our users... still a long way to go on the path to constraint management :-)